cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -163,22 +165,42 @@ public void replaceMetadata(MetadataScope scope, Metadata metadata,
         propertiesToSet = Maps.filterKeys(propertiesToSet, key -> !oldPropertiesToPreserve.contains(key));
       }
       // perform updates
-      mds.removeTags(entity, tagsToDelete);
-      mds.addTags(entity, tagsToAdd);
-      mds.removeProperties(entity, propertiesToDelete);
-      MetadataChange changed = mds.setProperties(entity, propertiesToSet);
+      MetadataChange changed = null;
+      if (!tagsToDelete.isEmpty()) {
+        changed = mds.removeTags(entity, tagsToDelete);
+      }
+      if (!tagsToAdd.isEmpty()) {
```

Comment:
I found an existing bug in addTags() if the set of tags to add is empty, then it removes the existing tags from the index. I can't figure out why that is happening without intense investigation. So for now I changed this code to avoid calling any mds method with an empty set or map. It is better anyway as it saves I/O in the case where no update is needed. 

Jira is filed for the addTags() bug and test cases added.